determine_needed_tags.sh

```diff
@@ -0,0 +1,147 @@
+  echo
+  echo "Run $0 -h|--help to show this usage information."
+}
+
+case $1 in
+  -h|--help)
+    showHelp
+    exit
+  ;;
+esac
+
+if [ "$#" -ne 2 ]; then
+  echo "Error: Missing or unexpected number of positional arguments. Expected: 2"
+  echo
+  showHelp
+  exit 1
+fi
+
+if [[ ${build_version} =~ ^([0-9]+)\.([0-9]+).?([0-9]*)$ ]]; then
+  build_major=${BASH_REMATCH[1]}
+  build_minor=${BASH_REMATCH[2]}
+  build_patch=${BASH_REMATCH[3]}
+else
+  echo "The provided version ${build_version} does not have the expected format."
+  exit 1
+fi
+
+if [ -z ${build_patch} ]; then 
+  build_patch=0
+fi
+
+tag_update_flag=0
+# possible values of the tag_update flag:
+# 0 = no additional tags to be set
+#     (this applies just in case of builds for older releases while a newer minor or minor.patch is already available)
+# 1 = add the major.minor tag
+#     (this applies, if the build is for is just for a new patch version)
+# 2 = add the tags major and major.minor
+#     (this applies if the build is for a version equal to or greater than the latest minor of a major)
+# 3 = add the tags latest in addition to the tags from #2
+#     (this applies if the build with the major, minor and patch version being equal to or greater than the currently latest version
+#     or if the build is for a completely new st2 component)
+
+# check if there are already images for the given component available at dockerhub
+dockerhub_registry_status_code=$(curl -sfL -o /dev/null -w "%{http_code}" https://registry.hub.docker.com/v1/repositories/stackstorm/${component}/tags)
+
+if [ ${dockerhub_registry_status_code} -eq 404 ]; then
+  # there is no repository stackstorm/${component} available at dockerhub -> this build is for a new st2 component
+  tag_update_flag=3
+  echo $tag_update_flag
+  exit
+elif [ ${dockerhub_registry_status_code} -ne 200 ]; then
+  echo "Unexpected HTTP statuscode for https://registry.hub.docker.com/v1/repositories/stackstorm/${component}/tags: ${dockerhub_registry_status_code}"
+  exit 1
+fi
+
+# dockerhub lists the tags in ascending order. 1st object = lowest tag; last object = highest tag or latest
+docker_tags_json=$(curl -s https://registry.hub.docker.com/v1/repositories/stackstorm/${component}/tags)
+readarray -t available_releases < <(echo $docker_tags_json | jq -r '.[] | select(.name | endswith("dev") | not).name' | sort)
+
+# sort returns the list with the highest tag or "latest" as last item
+# so change the value below to -2 when introducing the tag latest to get i.e. 3.3.0 instead of latest
+latest_release=${available_releases[-1]}
+latest_release_array=(${latest_release//\./ })
+
+latest_major=${latest_release_array[0]}
+latest_minor=${latest_release_array[1]}
+
+
+if [ ${build_version} == ${latest_release} ]; then
+  # building a release of the latest st2 version
+  tag_update_flag=3
+else
+  # building a release for a st2 version that does not match the latest version
+  if [ ${build_major} -le ${latest_major} ]; then
```

Comment:
Here is what I got when running the script locally: 
```
./determine_needed_tags.sh st2 3.0.0

./determine_needed_tags.sh: line 95: [: latest: integer expression expected
3
```

Reply:
Ofc, that shouldn't happen. I'll take a look at it tomorrow evening latest and try to provide a fix on the weekend. Will also see how this passed my own tests then.

Reply:
Oh ok, I see how this happened. It looks like there are images at Dockerhub that do already have the latest tag and this one is now provided as the last item in the list (and no longer the actual latest release i.e. 3.3.0). 

I'll address that shortly.

Reply:
Ok, this is fixed now. The main issue was the one I already described above: the script expected a release version like 3.3.0 and not `latest`. Now, `latest` as available release is also filtered and no longer returned by jq. 

The more important point is your 2nd one: 
```
./determine_needed_tags.sh: line 95: [: latest: integer expression expected
3
```
This was a glidge in the script due to the "wonderful" (sorry, don't find an appropriate word for it right now :) ) error handling in bash. 

What happend is that the if clause failed due to `latest` not being an integer and falling back to the else clause. I addressed this by changing: 
```
  if [ ${build_major} -le ${latest_major} ]; then # < this one failed
[...]
  else # < so we got here
    # building a release for a new major version
    tag_update_flag=3
  fi
```
to: 
```
  if [ ${build_major} -le ${latest_major} ]; then
[...]
  elif [ ${build_major} -gt ${latest_major} ]; then
    # building a release for a new major version
    tag_update_flag=3
  fi
```
So the worst thing that could happen now if parsing of a returned tag leads to an unwanted result is the error below: 
```
./determine_needed_tags.sh: line 95: [: latest: integer expression expected
./determine_needed_tags.sh: line 141: [: latest: integer expression expected
0
```

So the script returns 0 and no tag would be updated. I'll see if there is a more elegant solution to address this and even avoid (=handle) the 2 error messages.